README.md
# Grouped Representation of Interatomic Distances (GRID)

This package is designed to compute GRID descriptions of 
crystal structures and use them to train ML models, currently
based on properties extracted from the Materials Project.
In addition, it contains a number of tools for computing 
earth mover's distance (EMD) between distributions such 
as GRID or RDF, and using the resulting dissimilarities for
further calculations.

This code accompanies the following paper, which should be cited
if you use it for any future publications:

[Grouped Representation of Interatomic Distances as a Similarity Measure for Crystal Structures](https://doi.org/10.26434/chemrxiv-2022-9m4jh)





# Installation

The latest stable version of gridrdf can be installed using pip:

```
pip install gridrdf
```

If you are using conda, you may find it easier to create a new environment with the
required dependencies first, before installing gridrdf using pip:

```
conda create -n gridrdf_env -f environment.yml
conda activate gridrdf_env
pip install gridrdf
```


Alternatively, the most recent development version can be installed
by cloning the git repository, and then installing in 'development' mode:

```
git clone https://git.ecdf.ed.ac.uk/funcmatgroup/gridrdf.git
pip install -e gridrdf/
```

Using conda with this approach, you can install the dependencies from requirements.txt:

```
git clone https://git.ecdf.ed.ac.uk/funcmatgroup/gridrdf.git
conda env create -n gridrdf_env --file gridrdf/requirements.txt -c defaults -c conda-forge
conda activate gridrdf_env
pip install -e gridrdf
```

# Testing

Once downloaded or installed, it is recommended to test the code operates
correctly. Using a python terminal, navigate to the `gridrdf` directory and type

``` bash
python -m unittest discover -s tests
```

# Using the Code

All modules contained in gridrdf have documentation describing their
intended use, and are grouped into 'data preparation' (`gridrdf.data_prepare`),
'similarity calculation' (`gridrdf.earth_mover_distance`) and 'model training' (`gridrdf.train`) steps. 
Other utility modules are also included.

Submodules of gridrdf can be imported and used interactively in a python environment, but the main steps
outlined above can also be accessed as command line scripts by calling the module directly (--help will give 
more details of usage):

```
python -m gridrdf.MODULE_NAME --help
```

# Intended Workflow

To re-create the results presented in the publication of predicting
bulk modulus  using a kNN model and EMD dissimilarity, the procedure is as follows:

1. Import data from the materials project with calculated
   elastic moduli
   ``` python
   data = gridrdf.data_prepare.get_MP_bulk_modulus_data(APIkey)
   with open('MP_modulus.json') as f:
       gridrdf.json.dumps(data, f)
   ```
   
   NOTE: gridrdf currently relies on the legacy Materials Project API, so needs an old API KEY
   
2. Calculate GRID representation for each structure (generates GRID file for each structure)
   ``` python
    gridrdf.data_prepare.batch_rdf(data[:2],
                                   max_dist=10,
                                   bin_size = 0.1,
                                   method='kde',
                                   output_dir = './GRIDS',
                                   normalize=True
                                  )
   ```
   or from a terminal:
   
   ``` bash
   python -m gridrdf.data_prepare --data_source MP_modulus.json --output_dir ../GRIDS/ --tasks grid_rdf_kde
   ```
   
3. Remove any structures with fewer than 100 GRID shells
   ```python
   all_GRID = gridrdf.data_io.rdf_read_parallel(data, rdf_dir = './GRIDS/')
   for i, d in enumerate(data[:]):
       if len(all_GRID[i]) < 100:
           data.remove(d)
   with open('MP_subset.json', 'w') as f:
       json.dump(data, f, indent=1)
   ```   
   or from a terminal:
   ``` bash
   python -m gridrdf.data_prepare --data_source MP_modulus.json --output_dir ./GRIDS/ --tasks subset_grid_len --output_file MP_subset.json  
   ```
    
4. Filter structure with negative bulk moduli
   ``` python
   for d in data:
       if d['elasticity.K_VRH'] < 0:
           data.remove(d)
   ```
   or from a terminal:
   ``` bash
   python -m gridrdf.data_prepare --data_source MP_modulus.json --output_dir ./GRIDS/ --output_file MP_subset.json --tasks subset_property --prop_filter elasticity.K_VRH 0 np.inf
   ```
   
5. Filter elements with atomic number > Bi:
   ``` python
   # First, generate internal list of 78 elements (as gridrdf.composition.periodic_table_78)
   gridrdf.composition.element_indice()
   data = gridrdf.data_prepare.elements_selection(data, gridrdf.composition.periodic_table_78, mode='consist')
   ```
   
   NOTE: not currently implemented for command line script
    
Steps 2-5 can be combined into a single function call (similarly through terminal script by specifying tasks in order):

``` python
data_quick = gridrdf.data_prepare.main(data_source = './MP_modulus.json',
                                  tasks = ['subset_grid_len', 'subset_composition', 'subset_property'],
                                  output_dir = './GRIDS',
                                  output_file = 'subset.json',
                                  max_dist=10,
                                  min_grid_groups = 100,
                                  composition = {'elem': gridrdf.composition.periodic_table_78, 'type':'consist'},
                                  data_property = ('elasticity.K_VRH', 0, np.inf)
                                 )
```
    
    
6. Calculate pair-wise dissimilarity matrix between structures using EMD (time-consuming)
   ```
   similarity = gridrdf.earth_mover_distance.rdf_similarity_matrix(data, all_GRID, method='emd')
   similarity.to_csv('GRID_sim_whole_matrix.csv')
   ```
   or from a terminal:
   ```
   python -m gridrdf.earth_mover_distance --input_file MP_modulus.json --rdf_dir ./GRIDS/ --output_file GRID_sim --task rdf_similarity_matrix
   ```
   Note: The data can also be processed in smaller chunks using `indice` (or `--data_indice` as a script) to allow parallel-processing.
7. Use a simplified kNN model to predict bulk modulus
   ```
   K_data = np.array([ x['elasticity.K_VRH'] for x in data ])
   model = sklearn.neighbors.KNeighborsRegressor(n_neighbors=1, metric='precomputed')
   gridrdf.train.calc_obs_vs_pred_2D(funct = model,
                                     X_data = similarity,
                                     y_data = K_data,
                                     test_size = 0.2,
                                     outdir= './',
                                    )
   ```
   or from a terminal:
   ```
   python -m gridrdf.train --input_file MP_modulus.json --rdf_dir ./GRIDS/ --input_features distance_matrix --dist_matrix GRID_sim_whole_matrix.csv --out_dir ./ --funct knn_reg --target bulk_modulus --metrics emd --task obs_vs_pred
   ```
   
   
# Issues

If you have any questions, comments or problems with the code, please feel free to post them as issues [here](https://git.ecdf.ed.ac.uk/funcmatgroup/gridrdf/-/issues)! 
   


CHANGELOG.rst
=========
Changelog
=========

Version 0.3.0
=============

Major speed optimisations, particularly for computing EMD between GRID representations.

Backwards compatibility with previous versions is (almost) entirely maintained, but this version introduces
additional API methods. Future versions will likely use the newer array-based API approach, and deprecate
the older "list of dicts" approach.

Changes
-------

**EMD calculation**
- Numba-optimised calculation of EMD between GRID representations, both individually and computing
  entire EMD dissimilarity matrices. This makes particular use of cumulative GRID distributions internally.
  Additional features include memory-reduction techniques, for instance saving EMD results direct to disk
  (using e.g. h5py) without storing in memory.
      As an example, the calculation of a 12,000x12,000 EMD matrix based on 100 GRID shells takes approximately
      30 minutes on a desktop PC (Intel i5-8500, 6 cores) compared to ~30 days previously.
- Added initial ability to combine EMD between different GRID shells in different ways. Currently, a power-law 
  (shell_number**-n) is implemented to weight comparison equally between shells (n = 0) through to nearest neighbours
  dominating (n >= 1). A callable can also be provided to define the weights, but they must sum to n_shells.

**GRID generation**
- Major overhaul of GRID generation, so that Gaussian broadening and histogram binning are now handled using
  the same interface.
  This also fixes a bug where the KDE-broadened distances were not correctly integrated within
  each histogram bin, so narrow broadening oculd result in missing distribution. This is now improved to compute
  the KDE on a finer series of positions before later binning as desired, making it closer to a true integral.
- It is also now simpler to compute GRID representations without saving to disk, and collections of GRIDs can
  be saved/loaded as single files (e.g. npy or h5py).
- Initial implementation of a composition-based GRID and minor updates to composition handling.

**Other Changes**
- Improved testing, particularly of GRID generation and EMD computation
- Updated documentation
- Modified example scripts


Version 0.2.0
=============

- Changed to PyScaffold for development framework
- Added initial documentation, which is still in development
- A script to reproduce the results of the GRID publication is now
  available in the ``examples`` folder
- 


Version 0.1.3
=============

- Initial public release of gridrdf
- Functions modified to import as packages in the most part, and an initial bank of tests introduced
- Some optimisations to speed up computation and processing of GRIDs, for instance using multiprocessing


data_explore.py

import numpy as np
import pandas as pd
import json
import time
import gzip
import argparse
import os
from tqdm import tqdm
from pyemd import emd, emd_with_flow
from pymatgen import Structure
from pymatgen.analysis.local_env import CrystalNN
from pymatgen.symmetry.analyzer import SpacegroupAnalyzer 
try:
    from matminer.featurizers.structure import StructuralComplexity
except:
    print('matminer is not installed, cannot calculate structural complexity')

from extendRDF import rdf_histo, rdf_kde, get_rdf_and_atoms, shell_similarity
from composition import elements_count, bonding_type
from otherRDFs import origin_rdf_histo


def batch_rdf(data,
              max_dist=10,
              bin_size=0.1,
              method='bin',
              normalize=True, 
              gzip_file=False,
              output_dir = './'
              ):
    '''
    Read structures and output the extend RDF

    Args:
        data: input data from Materials Project
        max_dist: cut off distance of RDF
        method:
            bin: binned rdf, equal to the 'uniform' method in kde,  
                note that the 
            kde: use kernel density estimation to give a smooth curve
        gzip_file: in case RDF files take too much disk space, set 
                    this to true to gzip the files (not yet test)
    Return:

    '''
    for d in data:
        struct = Structure.from_str(d['cif'], fmt='cif')
        prim_cell_list = list(range(len(struct)))
        rdf_atoms = get_rdf_and_atoms(structure=struct, prim_cell_list=prim_cell_list, 
                                        max_dist=max_dist)
        
        if method == 'kde':
            rdf_bin = rdf_kde(rdf_atoms=rdf_atoms, max_dist=max_dist, bin_size=bin_size)
        elif method == 'bin':
            # this should be replaced by the general kde uniform method in the future
            rdf_bin = rdf_histo(rdf_atoms=rdf_atoms, max_dist=max_dist, bin_size=bin_size)
            if normalize:
                rdf_bin = rdf_bin / len(struct)
        else:
            print('This method is not supported in RDF calculation ')

        outfile = os.path.normpath(os.path.join(output_dir, d['task_id']))
        if gzip_file:
            with gzip.open(outfile+'.gz', 'w') as f:
                # not yet test, need test before use
                f.write(rdf_bin.tostring())
        else:
            np.savetxt(outfile, rdf_bin, delimiter=' ', fmt='%.3f')
        time.sleep(0.1)
    return


def batch_shell_similarity(all_rdf, method='append'):
    '''
    Calculate shell similarity for multiple RDFs
    
    Args:
        all_rdf: multiple RDFs with the same length
        method: how shell similarity is used in the X input
            append: append the similarity values after rdfs
            only: only use similarity values as the X input
    Return:
        RDF with shell similarity or similarity only
    '''
    rdf_shell_simi = []
    for rdf in all_rdf:
        if method == 'append':
            rdf_shell_simi.append(np.append(rdf, shell_similarity(rdf)))
        elif method == 'only':
            rdf_shell_simi.append(shell_similarity(rdf))
    return np.stack(rdf_shell_simi)


def rdf_trim(all_rdf, trim='minimum'):
    '''
    Trim the rdfs to the same length for kernel methods training

    Args:
        all_rdf: a list of np array (rdfs) with different length
        trim: must be one of 'no', 'minimum' or an integer
            no: no trim when the RDF already have the same length, 
            minimum: all rdf trim to the value of the smallest rdf
            integer value: if a value is given, all rdf longer than 
                this value will be trimmed, short than this value will  
                add 0.000 to the end
    Return:
        a two dimensional matrix, first dimension is the number of
        samples, and the second dimension is the flatten 1D rdf    
    '''
    rdf_lens = []
    for rdf in all_rdf:
        rdf_lens.append(len(rdf))

    if trim == 'minimum':
        min_len = np.array(rdf_lens).min()
        all_rdf = [ x[:min_len] for x in all_rdf]
    elif isinstance(trim, int):
        for i, rdf_len in enumerate(rdf_lens):
            if rdf_len < trim:
                b = trim - rdf_len
                if b != 0:
                    if len(all_rdf[0].shape) == 2:
                        nbins = len(all_rdf[i][0])
                        all_rdf[i] = np.append( all_rdf[i], [[0.0] * nbins] * b, axis=0 )
                    elif len(all_rdf[0].shape) == 1:
                        all_rdf[i] = np.append( all_rdf[i], [0.0] * b )
            else:
                all_rdf[i] = all_rdf[i][:trim]

#            print(len(all_rdf[i]))
    elif trim == 'none':
        pass 
    else:
        print('wrong value provided for trim') 

    return np.stack(all_rdf)


def rdf_flatten(all_rdf):
    '''
    If the rdf is not 1d, make it 1d for machine learning input
    This is done after trim so np.stack can be used

    Args:
        all_rdf: a 2D or 3D np array (rdfs) with the same length, 
            with the first dimension be number of samples
    Return:
        a 2D np array of rdfs
    '''
    if len(all_rdf[0].shape) == 2:
        all_rdf = [ x.flatten() for x in all_rdf]
    return np.stack(all_rdf)


def batch_lattice(data, method='matrix'):
    '''
    Calculate lattice parameters of each structure

    Args:
        data: input data from Materials Project
        method:
            abc: binned rdf, equal to the 'uniform' method in kde,  
                note that the 
            matrix: use kernel density estimation to give a smooth curve
    Return:
        lattice parameters
    '''
    all_lattice = []
    for d in data:
        struct = Structure.from_str(d['cif'], fmt='cif')
        if method == 'matrix':
            all_lattice.append(struct.lattice.matrix.reshape(9))
        elif method == 'abc':
            all_lattice.append(np.array(struct.lattice.abc + struct.lattice.angles))
    return np.stack(all_lattice)


def bond_length_statis(structure): 
    '''
    Calculated some bond length statistics values in the structure
    the nearest atoms pairs are determined using pymatgen CrystalNN module

    Currently the range of bond length has some problem for structure with
    no bonding, i.e. empty np array

    Args:
        structrue: pymatgen structure
    Return:
        the mean value and standard deviation and all the cation-anion
        bond length in the crystal structure
    '''
    nn = CrystalNN()
    bond_len = []
    for j in range(len(structure)): 
        for i in nn.get_nn_info(structure, j):       
            bond_len.append(structure[j].distance(structure[i['site_index']]))
    bond_len = np.array(bond_len)
    return bond_len.mean(), bond_len.std()#, bond_len.ptp()


def average_coordination(structure):
    '''
    Calculation of average coordination number over every site in a structure
    using Vorini method

    Args:
        structure: pymatgen structure
    Return:
        Average coordination number
    '''
    nn = CrystalNN()
    ave_coord_num = []
    for atom_site in range(len(structure)):
        ave_coord_num.append(len(nn.get_nn_info(structure, atom_site)))
    return np.array(ave_coord_num).mean()


def bond_stat_per_site(structure):
    '''
    Get the bond length standard deviation of each site in the unit cell, the average
    Get the coordination number of each site, then standard deviation

    Args:
        structure: pymatgen structure
    Return:
        Average of bond length std of each site
        Standard deviation of coordination number
    '''
    nn = CrystalNN()
    coord_num = []
    bond_len_std = []
    for atom_site in range(len(structure)):
        struct_nn = nn.get_nn_info(structure, atom_site)
        coord_num.append(len(struct_nn))
        bond_len = []
        for i in struct_nn:       
            bond_len.append(structure[atom_site].distance(structure[i['site_index']]))
        bond_len_std.append(np.array(bond_len).std())
    return round(np.array(bond_len_std).mean(), 3), \
            round(np.array(coord_num).std(), 3)


def num_of_shells(data, dir):
    '''
    Calculate the number of nearest neighbor shells in RDF for each compound,  
    some other properties are also given for data analysis purpose

    Args:
        data: the bulk modulus data and structure from Materials Project
        dir: the dir contains the calculated RDFs
    Return:
        a list of properties, will change according to requrest
    '''
    #sc = StructuralComplexity()
    results = []
    for d in data:
        struct = Structure.from_str(d['cif'], fmt='cif')
        #complex_atom, complex_cell = sc.featurize(struct)
        '''num_shell = sum(1 for line in open(dir + '/' + d['task_id']))

        with open(dir + '/' + d['task_id']) as f:
            rdf = np.loadtxt(f, delimiter=' ')
            num_rdf_bar = np.count_nonzero(rdf[:30])
        '''
        bond_mean, bond_std = bond_length_statis(struct)

        results.append(np.array([
            #d['elasticity.K_Voigt'],           # bulk modulus
            #num_shell,                          # number of RDF shells
            #num_rdf_bar,                        # non-zeors in the first 30 RDF shell
            #struct.volume/len(struct),          # volume per atom
            struct.volume,                      # volume
            struct.get_space_group_info()[1],   # space group number
            bond_mean,
            bond_std,
            #,     # average coordination nunmber
            #struct.density,                     # density in g/cm^3
            len(struct),                        # number of atoms
            len(struct.symbol_set),             # number of types of elements
            #sc.featurize(struct)[0],    # structural complexcity per atom
        ]))

    return np.stack(results)


def rdf_value_stat(data, dir):
    '''
    Remove all the 0 values in all RDF and put all other values in one list.  
    To get a statistics intuition of the RDF values distribution histogram.

    Args:
        dir: the dir contains the calculated RDFs
    Return:
        a list of all non-zero RDF values
    '''
    for d in data:
        struct = Structure.from_str(d['cif'], fmt='cif')
        len(struct)
        outfile = d['task_id']
    return


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Data explore',
                                    formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('--input_file', type=str, default='../MP_modulus.json',
                        help='the bulk modulus and structure from Materials Project')
    parser.add_argument('--output_file', type=str, default='rdf_similarity',
                        help='currently for rdf similarity')
    parser.add_argument('--rdf_dir', type=str, default='./',
                        help='dir has all the rdf files')
    parser.add_argument('--task', type=str, default='num_shell',
                        help='which property to be calculated: \n' +
                            '   num_shell: number of RDF shells and other propreties \n' +
                            '   extend_rdf_bin: binned extend RDF of all the CIFs \n' +
                            '   extend_rdf_kde: kernel density estimation RDF \n' + 
                            '   origin_rdf: calcualte vanilla RDF of all the CIFs \n' + 
                            '   composition: element-wise statistics of all compositions \n' +
                            '   bonding_type: \n' +
                            '   shell_similarity: \n' +
                            '   '
                      )
    parser.add_argument('--max_dist', type=float, default=10.0,
                        help='Cutoff distance of the RDF')

    args = parser.parse_args()
    input_file = args.input_file
    output_file = args.output_file
    rdf_dir = args.rdf_dir
    task = args.task

    max_dist = args.max_dist

    with open(input_file,'r') as f:
        data = json.load(f)

    if task == 'num_shell':
        results = num_of_shells(data=data, dir=rdf_dir)
        outfile = '../num_shell'
        np.savetxt(outfile, results, delimiter=' ', fmt='%.3f')
    elif task == 'extend_rdf_bin':
        batch_rdf(data, max_dist=max_dist, method='bin', output_dir = rdf_dir)
    elif task == 'extend_rdf_kde':
        batch_rdf(data, max_dist=max_dist, method='kde', output_dir = rdf_dir)
    elif task == 'origin_rdf':
        origin_rdf_histo(data, max_dist=max_dist)
    elif task == 'composition':
        elements_count(data)

    elif task == 'shell_similarity':
        for i, d in enumerate(data):
            if ( i % 100 ) == 0:
                print(i) 
            with open(rdf_dir + '/' + d['task_id']) as f:
                rdf = np.loadtxt(f, delimiter=' ')
                np.savetxt('../shell_similarity/' + d['task_id'], 
                            shell_similarity(rdf[:30]), 
                            delimiter=' ', fmt='%.3f')

    elif task == 'bonding_type':
        for d in data:
            struct = Structure.from_str(d['cif'], fmt='cif')
            d['bond_elem_list'], d['bond_num_list'] = bonding_type(struct)
        with open(input_file.replace('.json','_with_bond.json'), 'w') as f:
            json.dump(data, f, indent=1)

    elif task == 'bond_stat_per_site':
        for d in data:
            struct = Structure.from_str(d['cif'], fmt='cif')
            d['ave_bond_std'], d['coord_num_std'] = bond_stat_per_site(struct)
            d['num_sg_operation'] = len(SpacegroupAnalyzer(struct).get_space_group_operations())
        with open(input_file.replace('v7','v8'), 'w') as f:
            json.dump(data, f, indent=1)



    else:
        print('This task is not supported')

